devicetree-based device definitions and dependency representations

pabigot opened the pull request:
This PR uses the `DEVICE_DT_DEFINE()` infrastructure from #30196 to extract information about dependencies between devices from devicetree and encode it into the device structure using an integral type with unique values per device.

This is accomplished using the same two-pass model used for kernel objects: generate an initial object record using the raw data from the devicetree (handles are devicetree ordinals), then replace the handles with ones corresponding to the index of each device in the static device table.

The API supports recording additional dependencies on specific devices that do not have devicetree entries, such as the system clock.

A test is included, as well as an extension to the shell device module to display more information about devices, including initialization success and dependency data.

A commit from #30401 is replicated here.

pabigot committed:
device: inject device dependency information from devicetree nodes

Generate arrays of dependency information for each device.  If a
device definition is being constructed from devicetree these come from
the devicetree dependency information.  Additional dependencies may be
passed through using the DT_ macros.

Define flag values for device handles so we can partition the
dependency array into distinct sets, which include things it requires,
things it supports (may not be needed), and child nodes (not
implemented, may not be needed).

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: add post-process of elf file to manage device handles

Following the idiom used for system calls, add script support to read
the initial application binary to identify which devices are defined,
and to use their offset in the device array as their unique handle
rather than the externally-defined ordinal from devicetree.  The
device dependency arrays are updated to use these handles.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Compliance issues are related to multi-line macros used for defines, which cannot be enclosed in `do { } while` loops and need to have semicolon terminators.

galak commented:
Looks pretty good!

Would like to discuss how we can phase out `label` usage and under what conditions we need to support `device_get_binding` going forward.

carlescufi commented:
API Meeting 03 November 2020:

Discussion regarding the goal of "deprecating" the use of `DT_LABEL()` when invoking `dt_get_binding()`, and ultimately removing the need for a string in `device_get_binding()`.

tbursztyka commented on include/device.h:
```diff
@@ -220,6 +260,8 @@ struct device {
 	/** Address of the device instance private data */
 	void * const data;
+	/** optional pointer to handles associated with the device */
+	const device_handle_t *const handles;
```
Can't we get a more self-describing attribute name? (like childs, which is what these are no?)

And make the comment better describing the association. How is this associated? Or of course if the attribute is self describing, no need to tell more.

pabigot replied:
The array contains a sequence of sets of nodes.  Not all are child nodes; in the initial pass the first node is the self identifier and the reset are nodes it depends on (whether child or arbitrary, such as a chip-select GPIO).  If we need to know the node's parent device, it'd go here too.  I don't have a better name than "handles".

I've updated the description to make this more clear.

tbursztyka replied:
> The array contains a sequence of sets of nodes 

It's a simple pointer, so by itself it's a sequence of nodes. (sure, each node, translated to an actual device can point to another sequence etc... each device's handles attribute is a sub-tree then). 

> Not all are child nodes; in the initial pass the first node is the self identifier and the reset are nodes it depends on

So none are child nodes, since it cannot depend on its own child? Seems like it points to parent and/or dependency nodes only then. 
What do you mean by "the initial pass"? is that related to the state of the array before the elf script?

> If we need to know the node's parent device

How do you differentiate between a dependency node and its actual parent?

Maybe I missed it, but do you have a visual (ascii art, or picture etc..) of what the tree could look like? I think it will explain better.

pabigot replied:
> What do you mean by "the initial pass"? is that related to the state of the array before the elf script?

Yes.

> > The array contains a sequence of sets of nodes
> 
> It's a simple pointer, so by itself it's a sequence of nodes. (sure, each node, translated to an actual device can point to another sequence etc... each device's handles attribute is a sub-tree then).

No, it really is (a pointer to) a sequence of sets of nodes, where each set is a sequence of handles and the sets are separated by a special value, and the whole thing terminated by another special value.  I expect the specific sets encoded in the sequence to change over time as we need more or less information from devicetree.

> > Not all are child nodes; in the initial pass the first node is the self identifier and the reset are nodes it depends on
> 
> So none are child nodes, since it cannot depend on its own child? Seems like it points to parent and/or dependency nodes only then.

For now, yes.  In general, it's whatever device handles are relevant and definable at build time.

> > If we need to know the node's parent device
> 
> How do you differentiate between a dependency node and its actual parent?

Right now, we don't.  Nothing needs the parent as a distinct identifiable entity.

The dependencies are from the devicetree analysis, exposed by [this macro](https://docs.zephyrproject.org/latest/reference/devicetree/api.html?highlight=ordinals#c.DT_SUPPORTS_DEP_ORDS).  Basically it's the parent node and any node that is directly referenced by a property in the node.  That may change as we get more experience with what dependencies are needed, but however it's done they're all device references that will end up in this sequence.

> Maybe I missed it, but do you have a visual (ascii art, or picture etc..) of what the tree could look like? I think it will explain better.

No, I don't.  There's one in @nashif's power management architecture presentation, which I can't find right now and the link has expired from slack.  But the sorts of things that aren't parents include:
* gpio phandles, e.g. to the device that controls the signal for chip-select, reset, interrupt, ...
* parent bus devices
* clock or power domain devices, such as PWM timers and voltage/current regulators

The dependency relations define a directed acyclic graph, not a tree.

pabigot replied:
I had made this:
```
         * This encodes a sequence of sets of device handles that have
￼	 * some relationship to this node.  The individual sets are
￼	 * extracted with dedicated API, such as
￼	 * device_get_requires_handles().
```

I can't say what the sets will be because I don't know yet: this is a core capability for which applications will be identified over time.  But the values in this sequence are handles or non-handle marker values that delineate the ordered sequence of handle sets.

Could you suggest wording that describes this another way?

tbursztyka commented on include/device.h:
```diff
@@ -229,6 +271,78 @@ struct device {
+ * identify a device.
+ */
+static inline const struct device *
+device_from_handle(device_handle_t dh)
```
s/dh/dev_handle

tbursztyka commented on include/device.h:
```diff
@@ -229,6 +271,78 @@ struct device {
+ *
+ * @param dev the device for which a dependencies are desired.
+ *
+ * @param len pointer to a place to store the number of devices provided at the
```
s/len/count

pabigot replied:
`len` seems to be used more than `count` in Zephyr API that stores array lengths, but if you really want this I'll change it.

tbursztyka replied:
Then talk about array length in the documentation.

pabigot replied:
I like the existing text, so it's `count` now, and we have a sequence of count handles.

tbursztyka commented on include/device.h:
```diff
@@ -130,28 +152,49 @@ extern "C" {
 		      data_ptr, cfg_ptr, level, prio, api_ptr)
 
 #define DEVICE_DT_INIT(node_id, init_fn,				\
-		      data_ptr, cfg_ptr, level, prio, api_ptr)		\
```
Why removing api_ptr?

tbursztyka commented on scripts/gen_handles.py:
```diff
@@ -0,0 +1,290 @@
+        return False
+    return False
+
+DEVICE_HANDLE_SEP = -32768
```
Add a comment that internal dev handles are made of type int16_t and thus these markers make use of both limits.

tbursztyka commented:
@pabigot @nashif Besides my comment, I fully agree on the general approach.

pabigot commented on include/device.h:
```diff
@@ -125,33 +125,27 @@ extern "C" {
-	};								\
-	Z_INIT_ENTRY_DEFINE(_CONCAT(__device_, dev_name), init_fn,	\
-			    (&_CONCAT(__device_, dev_name)), level, prio)
+	Z_DEVICE_DEFINE(DT_INVALID_NODE, dev_name, drv_name, init_fn,	\
```
I don't understand this request.  It's DT-based devices that need more information passed in.

pabigot replied:
This has been addressed.

pabigot reviewed (commented):
@tbursztyka Thanks, dropping down to one macro makes it cleaner.  Most changes from your comments are in [this diff](https://github.com/zephyrproject-rtos/zephyr/compare/cde2c687e7bb9ebd67837708dc2f1bd782db3782..f8f57892d4ee3113fb9d6613e3351201b78dc039)

All comments addressed except one I don't understand ~, and renaming `len` to `count`, but I'll do that if you really want it~.  We need a style guide.

mbolivar-nordic commented:
Why are the devices still `static`?

I think the devicetree node identifiers are sufficiently unique that we should make the device structures global variables instead of static inside their drivers, at least when they are initialized from DT.

That would benefit boot time and footprint, as well as eliminate some boilerplate, in cases where one device depends on another.

For example, a sensor device that wants to get a pointer to its bus node would be able to do so at build time instead of boot time, as long as it has a devicetree node identifier for the bus -- which, in practice, it always does.

The devices themselves are const and userspace should be protecting them from undesired writes anyway. Are there some issues I'm not seeing other than name clashes (which again aren't an issue)?

pabigot commented:
> Why are the devices still `static`?

They were intended to be public.  Apparently the cases where I'm relying on getting the external reference directly from the node_id use a different define mechanism.  Or there's something else weird going on.

They do have to be static for the non-devicetree nodes.  So yes, this needs to get fixed.

mbolivar-nordic commented:
> So yes, this needs to get fixed.

I volunteer to update the bme280 driver test out resolving bus devices at build time whenever the fix is in. That's the driver I've mainly been using as a proving ground for new DT-based features.

mbolivar-nordic commented:
@pabigot thanks for the update. So far this is going pretty well. However, it's a bit of annoyance to have to add `extern const struct device DEVICE_DT_NAME_GET(node_id);` lines in the driver just to be able to use `DEVICE_DT_GET(node_id)`, especially because the `DEVICE_DT_GET()` macros are inside device instantiation macros, so I have to come up with per-instances helpers that look like this:

```
#define DECLARE_DT_DEVICE(node_id) \
	extern const struct device DEVICE_DT_NAME_GET(node_id)

#define MY_DEVICE(inst) \
	DECLARE_DT_DEVICE(something_I_care_about(inst)); \
	<config, data, DEVICE_DEFINE calls go here>
```

Would it be possible to generate a header with all the possible `extern const struct device` declarations in it already, that would be included by `<device.h>`? This could be done for every node_id in the devicetree, for example, even if some of those don't lead to real devices.

I think the convenience for device drivers outweighs the fact that some declarations are "false positives" not pointing at a real device.

pabigot commented:
> However, it's a bit of annoyance to have to add `extern const struct device DEVICE_DT_NAME_GET(node_id);`

That's what `DEVICE_DT_DECLARE(node_id);` was for.  I did mistakenly remove the `extern` from it in a recent fix, but it's back now, and should be enough to keep moving for now.

Yes, it should be possible to have `gen_handles` generate a header that contains the declaration for all devices that derive from devicetree-supporting drivers.  I'll work on that, but probably not until next week.

pabigot commented:
> Yes, it should be possible to have `gen_handles` generate a header that contains the declaration for all devices that derive from devicetree-supporting drivers.

Of course `gen_handles` is far too late for that, since the header has to be available when the sources are initially compiled, not just for the final link.  So the header would have to be generated before anything's compiled, probably by the standard devicetree processing along with whatever produces the final `zephyr.dts`.

@mbolivar that seems more in your area.  Could you do that?  You should be able to just generate a list of `DEVICE_DT_DECLARE()` instances.

mbolivar-nordic commented:
> Of course `gen_handles` is far too late for that, since the header has to be available when the sources are initially compiled, not just for the final link.

Right, that's what I'm talking about when I refer to false positives etc. Please take another look at my request. I anticipated this.



> @mbolivar-nordic that seems more in your area. Could you do that? You should be able to just generate a list of `DEVICE_DT_DECLARE()` instances.

I can take a crack at it if you don't have time, sure.

pabigot commented:
OK, looking again:
> so I have to come up with per-instance helpers that look like this:
> 
> ```
> #define DECLARE_DT_DEVICE(node_id) \
> 	extern const struct device DEVICE_DT_NAME_GET(node_id)
> 
> #define MY_DEVICE(inst) \
> 	DECLARE_DT_DEVICE(something_I_care_about(inst)); \
> 	<config, data, DEVICE_DEFINE calls go here>
> ```

`DECLARE_DT_DEVICE` already exists as `DEVICE_DT_DECLARE`.  I don't understand what `MY_DEVICE(inst)` is intended for; the declaration was all I needed for in-module use of this sort of thing, which you can see in the Nordic clock control and UART driver updates.

> I can take a crack at it if you don't have time, sure.

I won't have time until next week, and I think it really belongs more in code you own.  I'd guess it should go into a file generated alongside `devicetree_unfixed.h`.  AFAICT there's nothing in this PR that should need to change to add these declarations.

mbolivar-nordic commented:
This is the part I meant:

> Would it be possible to generate a header with all the possible extern const struct device declarations in it already, that would be included by <device.h>? This could be done for every node_id in the devicetree, for example, even if some of those don't lead to real devices.

mbolivar-nordic commented:
>  I'd guess it should go into a file generated alongside `devicetree_unfixed.h`. AFAICT there's nothing in this PR that should need to change to add these declarations.

That is a layering mistake IMO.

I'm talking about generating a header that looks like this:

```
extern const struct device __device_D_N;
extern const struct device __device_D_N_S_soc;
...
extern const struct device __device_DT_N_S_soc_S_i2c_40003000;
extern const struct device __device_DT_N_S_soc_S_i2c_4000XYZW;
...
```

Some of those declarations become real devices, some don't.

The entire generated header would get included by device.h.

This doesn't belong in devicetree_foo.h because that's underneath the device layer. Putting this content in devicetree_unfixed.h would be putting content from a higher layer in a lower layer.

mbolivar-nordic commented:
> I volunteer to update the bme280 driver to test out resolving bus devices at build time whenever the fix is in.

This is working now:
https://github.com/mbolivar-nordic/zephyr/tree/bme280-on-peter-device

I think the results look pretty decent, actually. I would like to see the same approach adopted in other drivers.

@avisconti what do you think of that for cutting down on boilerplate for "SPI2C" devices? See the `sensor: bme280: obtain device pointers directly` patch in particular.

@tbursztyka would you be open to adding the SPI helper macros in `[wip] drivers: spi: add helpers for initializing from DT` that this depends on to the SPI API?

pabigot commented:
> > I'd guess it should go into a file generated alongside `devicetree_unfixed.h`. AFAICT there's nothing in this PR that should need to change to add these declarations.
> 
> That is a layering mistake IMO.
> 
> I'm talking about generating a header that looks like this:
> 
> ```
> extern const struct device __device_D_N;
> extern const struct device __device_D_N_S_soc;
> ...
> extern const struct device __device_DT_N_S_soc_S_i2c_40003000;
> extern const struct device __device_DT_N_S_soc_S_i2c_4000XYZW;
> ...
> ```
> 
> Some of those declarations become real devices, some don't.
> 
> The entire generated header would get included by device.h.
> 
> This doesn't belong in devicetree_foo.h because that's underneath the device layer. Putting this content in devicetree_unfixed.h would be putting content from a higher layer in a lower layer.

That wasn't what I meant.  What I meant was that the new header must be available before anything gets compiled---just like `devicetree_unfixed.h`.  That's because what gets compiled needs to have those declarations available.  Which suggests it should be generated by the existing tooling that does pre-compile devicetree analysis, and placed in the same `build/zephyr/include/generated/` directory.  It seems reasonable to have `device.h` include it from there.

This PR doesn't need to touch that tooling, and what it does touch is run too late to be useful, so I don't think that capability needs to be part of this effort.

But if I'm still missing something let me know, and I can revisit this next week.

nvlsianpu reviewed (commented):
Looks good to me in general.
As I understand introduction of dependency representation it is a step towards better device management.

tbursztyka commented:
@mbolivar 
> @tbursztyka would you be open to adding the SPI helper macros in [wip] drivers: spi: add helpers for initializing from DT that this depends on to the SPI API?

Sure, I'll take a look next week

pabigot commented:
@tbursztyka Please revisit; I believe all your comments have been addressed.

carlescufi commented:
@pabigot 
>I can't say what the sets will be because I don't know yet: this is a core capability for which applications will be identified over time. 

Could you provide an initial guess of a couple of use-cases here? Just to have a rough idea, since the sets will be there there must be a very coarse concept idea for their use?

pabigot commented:
> Could you provide an initial guess at a couple of use-cases here? Just to have a rough idea, since the sets will be there there must be a very coarse concept idea for their use?

Values that have been proposed at various points:
* The handle for the parent device, if any.  This wouldn't need to be a set, it could be a value in an initial sequence of fixed length.
* The handles of child devices, if any.
* The handles for "devices" that this device depends on, if any.
* The handles for devices that depend on this device, if any.

Basically anything outside this device that can be uniquely identified by a 16-bit handle and that needs to be identified through a generic API.  The case we have now is devices (and possibly init objects) that a device depends on.

Use cases include:
* Walk all devices at init time, invoking their initializers only when everything they depend on has already been initialized.
* When entering or leaving different functional modes notify related devices of the introduction or removal of a dependency that affects the other device's functions.

tbursztyka commented on include/device.h:
```diff
@@ -297,6 +347,78 @@ struct device {
+}
+
+/**
+ * @brief Get the set of devices on which a given device depends
```
so, it might be populated of parent, direct dependency and child... which one are definitely not a dependency.

pabigot replied:
No.

What it returns is documented.  It would point to the set that represents the dependencies.  It doesn't have to return everything that's in handles.

tbursztyka commented:
@pabigot 
> Values that have been proposed at various points:
    The handle for the parent device, if any. This wouldn't need to be a set, it could be a value in an initial sequence of fixed length.
    The handles of child devices, if any.
    The handles for "devices" that this device depends on, if any.
    The handles for devices that depend on this device, if any.

How are you going to differentiate nodes against each other? (which one is a dependency, a parent or a child?)

And why having both dependencies and child in each device's private struct?
Having both is indeed going to make things much faster but still the graph isn't going to be that big (a thousand of device is going to be very rare), so how did you decide between memory footprint vs graph traversing optimization?

Also, see my comment about device_get_requires_handles().

Seems to me the whole require a bit more thoughts.

pabigot commented on include/device.h:
```diff
@@ -215,6 +347,78 @@ struct device {
+		while ((rv[i] != DEVICE_HANDLE_ENDS)
+		       && (rv[i] != DEVICE_HANDLE_SEP)) {
+			++i;
+		}
```
@tbursztyka What you see above is how you extract the first set.   Does that answer your question in https://github.com/zephyrproject-rtos/zephyr/pull/29644#issuecomment-727814647 about how the sets are distinguished?

galak commented:
dev-review (Nov 19, 2020):
* suggestion to split out the PR to have the DT macro additions and driver changes to allow that to get merged and open the ability for other drivers to get converted
* create a test that utilizes the dependency code

pabigot commented:
Back to draft until  #30196 gets in.

pabigot committed:
shell: device_service: add dependency information

Refactor the output of device list to use standard API to retrieve the
list of devices, and to always display a status rather than hiding
disabled/failed devices.

Add API to associate a distinct identifier with any "device" that does
not have a name.

Where a device has requires dependencies display the devices on which
it depends.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
dts: bindings: test: complete bindings for gpio expander

These need to be gpio controllers and have the necessary specifier
properties for use in gpio specifiers.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: devicetree: add test for devicetree node data

Verify that device dependencies are encoded into and retrievable from
the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Below is the output from the shell device module when built for the `thing52_nrf52832` with `CONFIG_CCS811=y`.  This is the most complex set of dependencies available until #30467 gets underway.
```
uart:~$ device list
devices:
- CLOCK (READY)
- UART_0 (READY)
- sys_clock (READY)
- GPIO_0 (READY)
- [0x2000029c] (READY)
- I2C_1 (READY)
- I2C_0 (READY)
- GPIO_P0 (READY)
  req: vdd-pwr-ctrl
  req: I2C_0
- ccs-pwr-ctrl (READY)
  req: GPIO_P0
- mic-pwr-ctrl (READY)
  req: GPIO_P0
- mpu-pwr-ctrl (READY)
  req: GPIO_P0
- spk-pwr-ctrl (READY)
  req: GPIO_0
- vdd-pwr-ctrl (READY)
  req: GPIO_0
- [0x20000350] (READY)
- CCS811 (READY)
  req: ccs-pwr-ctrl
  req: GPIO_P0
  req: I2C_0
  req: GPIO_0
uart:~$ 
```

nashif commented:
> - CCS811 (READY)
>   req: ccs-pwr-ctrl
>   req: GPIO_P0
>   req: I2C_0
>   req: GPIO_0
> uart:~$

I would replace req with `requires` to make it more clear

pabigot commented:
> I would replace req with `requires` to make it more clear

Done.  For the record: the build failures appear to be qemu_x86_64 on kernel/fifo/fifo_api, and the tests pass when built directly, but not under twister.  I don't have a solution for that.

mbolivar-nordic commented on include/device.h:
```diff
@@ -258,6 +286,7 @@ extern "C" {
 #define DEVICE_DT_DECLARE(node_id)			\
 	extern const struct device DEVICE_DT_NAME_GET(node_id)
 
+
```
?

mbolivar-nordic commented on include/device.h:
```diff
@@ -323,9 +369,6 @@ struct device_pm {
 	struct k_poll_signal signal;
 };
 
-/**
- * @brief Runtime device structure (in memory) per driver instance
- */
```
?

pabigot replied:
The original comment was wrong (this isn't mutable as implied), restored and corrected.

mbolivar-nordic commented on scripts/gen_handles.py:
```diff
@@ -0,0 +1,291 @@
+    global args
+
+    parser = argparse.ArgumentParser(
+        description=__doc__,
```
Did you mean to write a docstring here?

pabigot replied:
Didn't mean to do anything but copy from the existing scripts, but made this do something and fixed the other leftovers too, thanks.

mbolivar-nordic commented on scripts/gen_handles.py:
```diff
@@ -0,0 +1,291 @@
+        return struct.unpack(format, data)
+
+def process_device_symbol(sym, section, elf, edt):
+    if not sym.name.startswith("__device_"):
+        return False
+    return False
```
This looks wrong, but the function is unused. Just leftovers?

mbolivar-nordic commented on scripts/gen_handles.py:
```diff
@@ -0,0 +1,291 @@
+def main():
+    parse_args()
+
+    assert args.kernel, "--kernel ELF required for --gperf-output"
```
Is `--gperf` copy/pasta?

mbolivar-nordic reviewed (commented):
A few cosmetic nitpicks I wanted to get out of the way while digging into this again.

nashif commented on include/device.h:
```diff
@@ -344,6 +397,78 @@ struct device {
+ * @return the handle for the device.
+ */
+static inline device_handle_t
+device_get_handle(const struct device *dev)
```
`device_handle_get()` just like we have `DEVICE_NAME_GET` etc

nashif commented on include/device.h:
```diff
@@ -344,6 +397,78 @@ struct device {
+ * dev_handle does not identify a device.
+ */
+static inline const struct device *
+device_from_handle(device_handle_t dev_handle)
```
`device_from_handle_get()` maybe

pabigot replied:
`device from handle get` is pretty contorted syntax. But let's see what others think.

mbolivar-nordic replied:
DT APIs use "by", e.g. `DT_PHA_BY_IDX()`, `DT_PHA_BY_NAME()`, which could point to `device_get_by_handle()` or `device_by_handle()`.

That still looks weird to my US English eyes, though, in a way that `device_get_from_handle()` does not.

nashif commented on include/device.h:
```diff
@@ -344,6 +397,78 @@ struct device {
+/**
+ * @brief Get the set of devices on which a given device depends
+ *
+ * @param dev the device for which a dependencies are desired.
```
for which dependencies are desired

nashif commented on include/device.h:
```diff
@@ -725,18 +850,58 @@ static inline int device_pm_put_sync(const struct device *dev) { return -ENOTSUP
+ * dependency data.  If the object doesn't come from a devicetree
+ * node, use dev_name.
+ */
+#define Z_DEVICE_HANDLE_NAME(node_id, dev_name)				\
```
if all of those macros are just going to be used in this file, I would just use _DEVICE_HANDLE_NAME etc.
This Z_ thing is currently very ambiguous and was introduced for internal private APIs across multiple files in the same modules, but IMO should be omitted for helper functions and  macros inside the same file.

pabigot replied:
Those are *not* public API, they're an implementation detail of the public API that does use `DEVICE_HANDLE`.  They're globally visible and not to be used by anybody other than the implementation so I really think they should be namespaced.

tejlmand commented on CMakeLists.txt:
```diff
@@ -748,6 +748,19 @@ if(CONFIG_GEN_ISR_TABLES)
+  --kernel $<TARGET_FILE:${ZEPHYR_PREBUILT_EXECUTABLE}>
+  DEPENDS ${ZEPHYR_PREBUILT_EXECUTABLE}
+  )
+set_property(GLOBAL APPEND PROPERTY GENERATED_KERNEL_SOURCE_FILES dev_handles.c)
```
this actually has the side-effect that there will always be two stage linking.

Today, if `CONFIG_GEN_ISR_TABLES=n`
https://github.com/zephyrproject-rtos/zephyr/blob/201a8d29cdb2f856f7f889f9a75d3a463e3491ef/CMakeLists.txt#L716-L749

and `CONFIG_USERSPACE=n`
https://github.com/zephyrproject-rtos/zephyr/blob/201a8d29cdb2f856f7f889f9a75d3a463e3491ef/CMakeLists.txt#L756-L894

then the there is only a single linking stage, and the pre-built executable == final executable.
https://github.com/zephyrproject-rtos/zephyr/blob/201a8d29cdb2f856f7f889f9a75d3a463e3491ef/CMakeLists.txt#L1020-L1023

This PR changes that by always requiring two times linking.

@daor-oti @mped-oticon could you provide feedback on whether this approach impacts custom toolchains.
If I recall correctly, your build system disables `USERSPACE` and `GEN_ISR_TABLES` and thus you are not postprocessing the zephyr pre-built elf file.

daor-oti replied:
@tejlmand Correct, we have disabled `USERSPACE` and `GEN_ISR_TABLES` and thus run with a single stage linking. The overhead with the two stage linking would be undesirable in our case, so perhaps it could be configurable, whether or not to run `gen_handles.py`?

pabigot replied:
> it could be configurable, whether or not to run `gen_handles.py`?

No, device dependency information is critical for power management and safe initialization order.  If you cannot tolerate a second link operation when building then I'll have to move to mapping between device handles and actual devices by runtime code rather than an O(1) lookup.

daor-oti replied:
@pabigot Ok, it was just a thought that maybe it could be disabled, but it cannot. Unless others have a problem with the extra linking step and the small increase in build time it adds, I do not think at I can justify the rework and runtime overhead for moving to runtime code, so do not change it on just my account.

tejlmand reviewed (changes requested):
I have a little concern on always requiring a two stage linking.

Today, it is possible to build Zephyr and just link once.
Always post process zephyr.elf could make it harder to use Zephyr with custom toolchains that does things differently.

Are there no way we can obtain the needed information before linking stage ?

> This is accomplished using the same two-pass model used for kernel objects: 

Please note, that the `two-pass model used for kernel objects` is an optional step in current design, whereas this PR introduces a mandatory two-pass model.

pabigot commented:
> > This is accomplished using the same two-pass model used for kernel objects:
> 
> Please note, that the `two-pass model used for kernel objects` is an optional step in current design, whereas this PR introduces a mandatory two-pass model.
>
> Are there no way we can obtain the needed information before linking stage ?

Yes.  We could add infrastructure to do the runtime translation between the handle defined by the devicetree content and the handle to be used at runtime.  This would introduce two distinct handle values at runtime, which would be confusing, increase either flash or RAM usage, and impose an O(n) lookup each time a handle has to be converted to a device instance, unless it turned out to be possible to cache the lookup somehow.

I don't see a problem with two-stage linking, so I went with the simpler solution.

pabigot commented:
Please revisit this and provide/update your reviews: I've responded to the previous questions and have no changes planned absent further feedback.

pabigot commented:
@tejlmand @nashif https://github.com/zephyrproject-rtos/zephyr/pull/29644#issuecomment-756337154

pabigot commented:
Please note that all compliance failures in this are false positives.

tejlmand reviewed (approved):
lgtm

galak commented on scripts/gen_handles.py:
```diff
@@ -0,0 +1,306 @@
+# These match the corresponding constants in <device.h>
+DEVICE_HANDLE_SEP = -32768
+DEVICE_HANDLE_ENDS = 32767
+
```
Can we use the symbolic names from device.h instead?  So the code-gen looks like:

... `= { 5, DEVICE_HANDLE_ENDS, DEVICE_HANDLE_ENDS, DEVICE_HANDLE_ENDS, DEVICE_HANDLE_ENDS};`

pabigot replied:
I *think* so.  I recall I didn't because of a dependency cycle in include files, but that may have been long ago.  I'll give it a shot.

pabigot replied:
Done.

nashif commented:
@galak need you review here.
